### PR TITLE
[FIX] account: round epd vals

### PR DESCRIPTION
**Pull request by agr-odoo**
Set "Cash Discount Tax Reduction" to "Always"
Crete the following taxes:
- fixed136:
  - Tax Computation: Fixed
  - Amount: 0.1360
  - Included in price: False
  - Affect Base of Subsequent Taxes: True
- fixed176:
  - Tax Computation: Fixed
  - Amount: 0.1760
  - Included in price: False
  - Affect Base of Subsequent Taxes: True
- 6%:
  - Tax Computation: Percentage of Price
  - Amount: 6
  - Included in price: False
  - Base Affected by Previous Taxes: True
- 12%:
  - Tax Computation: Percentage of Price
  - Amount: 12
  - Included in price: False
  - Base Affected by Previous Taxes: True
 
Create the following products:
- A:
  - Sales price: 116.00
  - Customer taxes: 6%
- B:
  - Sales price: 0.91
  - Customer taxes: 6%
- C:
  - Sales price: 194.21
  - Customer taxes: fixed136, 12%
- D:
  - Sales price: 31.46
  - Customer taxes: fixed176, 12% 
 
Create a Payment Term "1/7 Net 30" with 1% discount if paid within 7 days
Create an invoice:
- Add payment term "1/7 Net 30"
- Create the lines
  - product A, quantity 6
  - product B, quantity 350
  - product C, quantity 1
  - product D, quantity 5 
 
Save

Issue: An error will block the action
"""
The move (Draft Invoice (* 44)) is not balanced.
The total of debits equals $1,483.87 and the total of credits equals $1,483.86.
You might want to specify a default account on journal "Customer Invoices" to automatically balance each move.
"""

This occurs because we compute the balance amounts for epd without rounding. Then in `needed` we sum up the amounts, ending up with 3 epd lines, having respectively the following balance:
- 10.145 -> 10.15 after rounding
- -13.6601 -> 13.66 after rounding
- 3.5151 -> 3.52 after rounding

So, because of the rounding, we will have 1 extra cent in debit amount and the invoice will result unbalanced

Fixing by rounding the partial epd amounts

opw-3327069

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by agr-odoo** [comment]
@william-andre

**Comment by william-andre** [comment]
Doesn't this PR fix your issue?
https://github.com/odoo/odoo/pull/124517

**Comment by agr-odoo** [comment]
nice PR but ineffective for this issue

**Comment by agr-odoo** [comment]
@william-andre 
https://github.com/odoo/odoo/pull/124517 does not work for this issue because it affect the rounding after the `needed` computation is done, which is already too late for this case
These are the `epd_needed_vals`  values in the mentioned use case
```
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 6.96, 'balance': 6.96, 'price_subtotal': -6.96}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': -6.96, 'balance': -6.96, 'price_subtotal': 6.96, 'tax_ids': [(<Command.CLEAR: 5>, 0, 0)]}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 3.185, 'balance': 3.185, 'price_subtotal': -3.185}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': -3.185, 'balance': -3.185, 'price_subtotal': 3.185, 'tax_ids': [(<Command.CLEAR: 5>, 0, 0)]}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 1.9421000000000002, 'balance': 1.9421000000000002, 'price_subtotal': -1.9421000000000002}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': -1.9421000000000002, 'balance': -1.9421000000000002, 'price_subtotal': 1.9421000000000002, 'tax_ids': [(<Command.CLEAR: 5>, 0, 0)]}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 1.5730000000000002, 'balance': 1.5730000000000002, 'price_subtotal': -1.5730000000000002}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': -1.5730000000000002, 'balance': -1.5730000000000002, 'price_subtotal': 1.5730000000000002, 'tax_ids': [(<Command.CLEAR: 5>, 0, 0)]}
```
And this is the result of `needed` (only values)
```
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 10.145, 'balance': 10.145, 'price_subtotal': -10.145}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': -13.6601, 'balance': -13.6601, 'price_subtotal': 13.6601, 'tax_ids': [(<Command.CLEAR: 5>, 0, 0)]}
{'name': 'Early Payment Discount (1.0%)', 'amount_currency': 3.5151000000000003, 'balance': 3.5151000000000003, 'price_subtotal': -3.5151000000000003}
```
After the needed computation we already have wrong values, as already mentioned those balances will round to
10.145 -> 10.15
13.6601 -> 13.66 
3.5151 -> 3.52
creating 1 extra cent in debit/credit

**Comment by agr-odoo** [comment]
@william-andre

File: addons/account/tests/test_early_payment_discount.py [hunk 1]
```diff
@@ -728,3 +728,59 @@ def test_mixed_epd_with_tax_deleted_line(self):
...
+                (116, 6, tax_6),
+                (0.91, 350, tax_6),
+                (194.21, 1, tax_136 | tax_12),
+                (31.46, 5, tax_176 | tax_12)
+            ]:
+                with invoice_form.invoice_line_ids.new() as line_form:
+                    line_form.price_unit = price_unit
+                    line_form.quantity = quantity
+                    for tax in taxes:
+                        line_form.tax_ids.add(tax)
```
Comment (smetl) [inline addons/account/tests/test_early_payment_discount.py:783]: Use a create instead plz

File: addons/account/tests/test_early_payment_discount.py [hunk 2]
```diff
@@ -728,3 +728,59 @@ def test_mixed_epd_with_tax_deleted_line(self):
...
+
+        # If invoice is not balanced the following line will raise
+        invoice_form.save()
```
Comment (smetl) [inline addons/account/tests/test_early_payment_discount.py:786]: the save is already triggered when exiting the context manager so this is useless fyi

**Review by smetl — APPROVED** [review]
@robodoo r+

**Comment by william-andre** [comment]
I'd like to understand why #124517 doesn't fix this though. It might mean that is it not working as intended.

**Comment by agr-odoo** [comment]
Imho https://github.com/odoo/odoo/pull/124517 works as expected, it just does not handle this case
Before this fix we were computing `epd_needed_vals` in a 'round globally' style (take a look at https://github.com/odoo/odoo/pull/126102#issuecomment-1621724585 for some example numbers).
In rounding each epd line amount separately, we introduced an extra cent of rounding error. https://github.com/odoo/odoo/pull/124517 couldn't make a difference, as the values were already wrong
The rounding error does not occur when computing the epd vals in 'round per line' style

However I wonder if this is really fixing the error and not just moving it somewhere else, perhaps a better solution is to check and eventually fix balance of computed epd vals

**Pull request closed without merging**